### PR TITLE
[FIX] website: fix the bug when deleting a page from the page manager

### PR DESCRIPTION
**Pull request by loco-odoo**
Steps to reproduce the bug:
- Install the "Surveys" application.
- Go to the "Website" application.
- Go in the page manager ("Site" > "Pages").
- Select a page (typically "Home").
- Try to delete it.
- Error of type "'survey.survey' object has no attributes 'name'".

The error comes from the fact that records do not necessarily have the
attribute called "name". To fix this error, this commit uses
"display_name" that is defined in all models.

If you put yourself in dev mode and you try to delete a page from the page manager,
another bug of type "Got duplicate key in t-foreach" appears. This
bug comes from the fact that "dependency_value" is an array. Due to
that, the current item of the "t-foreach" iteration is the current
value. The bug is fixed by ensuring that the value of "t-key" is the
current value index and not the value itself.

task-3082345
opw-3100483
opw-3111671

### CODE REVIEW

**Comment by loco-odoo** [comment]
The first version of the code to solve this problem used the ```getattr()``` function (```getattr(rec, 'name', f'{model_name} {rec.id}')```) in order to add a default value when the record does not have the ```name``` attribute. However, it appeared that this function leads to security issue.

File: addons/website/models/website.py
Comment (Guillaume-gdi) [inline addons/website/models/website.py:852]: Code looks good :+1: 
I just see in task description that you also have to add a test tour for that
image

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
Here is a small review.
Same as your other PR, the branch name should finish with `-loco`. (do not recreate a PR)

File: addons/website/models/website.py
Comment (loco-odoo) [inline addons/website/models/website.py:824]: To avoid this error:

owl_error

File: addons/web/static/lib/owl/owl.js
Comment (loco-odoo) [inline addons/web/static/lib/owl/owl.js:4268]: To avoid this error:
owl_error

File: addons/website/models/website.py
Comment (loco-odoo) [inline addons/website/models/website.py:824]: Same reason as before: To avoid the Owl error.

File: addons/website/static/src/components/dialog/page_properties.xml
Comment (loco-odoo) [inline addons/website/static/src/components/dialog/page_properties.xml:36]: Because "dependency_value" is an array, the current item is the current value so "t-key" has to be the index of it.

File: addons/website/models/website.py
Comment (rdeodoo) [inline addons/website/models/website.py:862]: Keep the initial `for` loop format, on top of being easier to read, it will (and should) be a one line fix.
> The first version of this fix used the getattr() [..] leads to security issue.

What do you mean by security issue? If you mean that `getattr` triggers a ci security, it doesn't mean that you should not use it. (You could bypass that with :point_down: but there is no point)
```py
'ok' if 'name' in self else 'ko'
```

`getattr` is exactly what you want instead of
```py
if hasattr(rec, 'name'):
    record_name = rec.name
else:
    record_name = f'{model_name} {rec.id}'
```
->
```py
record_name = getattr(rec, 'name', f'{model_name} {rec.id}')
```

File: addons/website/tests/__init__.py
Comment (rdeodoo) [inline addons/website/tests/__init__.py:16]: a b c d e

File: addons/website/tests/test_page_manager.py
Comment (rdeodoo) [inline addons/website/tests/test_page_manager.py:6]: Just to be sure, you have a python linter enabled and disabled the E302 rule (as per the officiel doc so it's fine even if I prefer to keep it myself)?

File: addons/website/static/tests/tours/page_manager.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/page_manager.js:5]: `deleteSelectedPage`
Be sure your linter is warning you about this

File: addons/website/static/tests/tours/page_manager.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/page_manager.js:7]: Nitpicking (don't change it, it will be a waste of time) but generally `Click on`, `Pick` seems weird

File: addons/website/static/tests/tours/page_manager.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/page_manager.js:48]: You could extract `tr:contains("Home"):has(td[name="website_id"]:empty)` to a variable to avoid duplicating it, it will help if this need to be changed/fixed later (generally avoid duplicate especially when long/complicated stuff like here the selector)

**Review by rdeodoo — CHANGES_REQUESTED** [review]
> This first fix done, an other bug of type "Got duplicate key in
t-foreach" appeared. This bug comes from the fact that
"dependency_value" is an array. Due to that, the current item of the
"t-foreach" iteration is the current value. The bug is so fixed by
ensuring that the value of "t-key" is the current value index and not
the value itself.

Not sure to understand why you need this fix "once you have fixed the survey case", while it works well without survey installed, could you explain?

**Comment by loco-odoo** [comment]
@rdeodoo 
> Not sure to understand why you need this fix "once you have fixed the survey case", while it works well without survey installed, could you explain?

You are right, I will re-write my commit message, it leads to confusion.  In fact, the error "Got duplicate key in
t-foreach" also appears without "survey" installed if you are in dev mode. So it is a another bug that is not related to the installation of "survey".

**Comment by rdeodoo** [comment]
Oh, perfect then, you can probably make a separated commit for that. In it, maybe worth mentioning https://github.com/odoo/odoo/commit/0d1c263298ca6724a6fe6c31e2a2d3221c5e8c75 (and explain the same thing)
Also https://github.com/odoo/odoo/pull/104425#discussion_r1008096429

File: addons/website/tests/test_page_manager.py
Comment (loco-odoo) [reply addons/website/tests/test_page_manager.py:6]: Yes, the E302 rule is disabled in my python linter (as the E301 and the E501)

File: addons/website/tests/__init__.py
Comment (rdeodoo) [inline addons/website/tests/__init__.py]: Almost :p (`test_page` before)

File: addons/website/static/tests/tours/page_manager.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/page_manager.js:14]: You can omit those `click` `run`, it's the default behavior

File: addons/website/static/tests/tours/page_manager.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/page_manager.js:29]: Add a blank line before your `registerWebsitePreviewTour` it will ease reading the file

**Review by rdeodoo — APPROVED** [review]
The commit you mentioned was the wrong one, from odoo-dev repo or something
You can see it when you access the commit (which was https://github.com/odoo/odoo/commit/3dfed47330b0ea355c4ad2abdf483e94dd3afab4):
image

Also, please use commit full link with markdown instead of the commit hash, so it's clickable from shell / IDE / whatever :+1: 

(I modified it alongside this quick [diff](https://github.com/odoo/odoo/compare/1f77ae8cd904bdf368dcc99b17079d415b789684..e607399b7138594b2e9e20590a386ab0a51bb995))

**Comment by rdeodoo** [comment]
@odoo/rd-security Hello, could we have your approval here about the `getattr` usage?

File: addons/website/models/website.py
Comment (mart-e) [inline addons/website/models/website.py:854]: Why not `display_name`? or
```suggestion
                    'record_name': rec.name if 'name' in rec else f'{model_name} {rec.id}',
```

Comment (rdeodoo) [reply addons/website/models/website.py]: We changed for that, thank you

**Review by rdeodoo — APPROVED** [review]
@robodoo r+ rebase-ff

File: addons/website/static/tests/tours/page_manager.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/page_manager.js:19]: Here I guess

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Actually, the tour is taking too long to load the dependencies, the RPC on full runbot is taking ~8/10 sec

It failed 2 times on the staging: 
image
https://runbot.odoo.com/runbot/build/22491735
http://runbot122.odoo.com/runbot/static/build/22491735-16-0/tests/22491735-16-0-all/screenshots/sc_20230105_170239_908921_TestWebsitePageManager.png

```
2023-01-05 17:02:01,308 13 INFO 22491735-16-0-all odoo.addons.website.tests.test_page_manager.TestWebsitePageManager.browser: Tour website_page_manager: step 'Pick Delete (trigger: .o-dropdown--menu span:contains("Delete"))' succeeded 
2023-01-05 17:02:10,331 13 INFO 22491735-16-0-all werkzeug: 127.0.0.1 - - [05/Jan/2023 17:02:10] "POST /web/dataset/call_kw/website/search_url_dependencies HTTP/1.1" 200 - 3966 0.841 8.185
2023-01-05 17:02:11,308 13 INFO 22491735-16-0-all odoo.addons.website.tests.test_page_manager.TestWebsitePageManager.browser: _check_for_tooltip 
```

The step to load the dependencies is taking 9 seconds which is actually making the 10 sec default timeout on step to trigger (as there is extra time here and there making it going over 10 sec)

Could you increase the timeout of the 2 steps loading the dependencies?

**Review by rdeodoo — APPROVED** [review]
@robodoo r+

Thank you

- Don't hesitate to build the enterprise branch in such situations, to simulate the testing that will be done in the staging and have more data that might make your test fail
You can do it from the build page, see:
image
- We can ask for multi builds when race condition are involved, it will create a new sub-build that will run the problematic test only X times, it's fine here, the error was obvious and the fix is fine
- Once it already went from draft to "ready for review", you don't have to move back a PR to draft every time you need to fix something after a review, unless you prefer to do it for your internal management of your PR but it's useless otherwise as:
  - it already pinged the code owners (generally if it pinged half he company and there is actually still some job to be done that wasn't planed we still move it back to draft to signal them it's not yet time to review but it's not the case here)
  - we internally don't rely on the PR state to know if it should be reviewed or not 
tl;dr: don't feel obligated to move PR back to draft
image

**Pull request closed without merging**
